packages/flutter_tools/lib/src/hot.dart

```diff
@@ -0,0 +1,470 @@
+    devFSStatus.stop(showElapsedTime: true);
+    printStatus('Synced ${getSizeAsMB(_devFS.bytes)} MB');
+    return true;
+  }
+
+  Future<Null> _cleanupDevFS() async {
+    if (_devFS != null) {
+      // Cleanup the devFS.
+      await _devFS.destroy();
+    }
+    _devFS = null;
+  }
+
+  Future<Null> _launchFromDevFS(ApplicationPackage package,
+                                String mainScript) async {
+    String entryPath = path.relative(mainScript, from: _devFSProjectRootPath);
+    String deviceEntryPath =
+        _devFS.baseUri.resolve(entryPath).toFilePath();
+    String devicePackagesPath =
+        _devFS.baseUri.resolve('.packages').toFilePath();
+    String deviceAssetsDirectoryPath =
+        _devFS.baseUri.resolve('build/flx').toFilePath();
+    String viewId = await observatory.getFirstViewId();
+    await observatory.runInView(viewId,
+                                deviceEntryPath,
+                                devicePackagesPath,
+                                deviceAssetsDirectoryPath);
+  }
+
+  Future<Null> _coldStartFromSources() async {
+    if (_devFS != null)
+      await _updateDevFS();
+    Status restartStatus = logger.startProgress('Cold restart...');
+    await _launchFromDevFS(_package, _mainPath);
+    restartStatus.stop(showElapsedTime: true);
+  }
+
+  Future<bool> _reloadSources() async {
+    if (observatory.firstIsolateId == null)
+      throw 'Application isolate not found';
+    if (_devFS != null)
+      await _updateDevFS();
+    Status reloadStatus = logger.startProgress('Performing hot reload');
+    try {
+      await observatory.reloadSources(observatory.firstIsolateId);
+    } catch (errorMessage) {
+      reloadStatus.stop(showElapsedTime: true);
+      printError('Hot reload was rejected:\n$errorMessage');
+      return false;
+    }
+    reloadStatus.stop(showElapsedTime: true);
+    Status reassembleStatus =
+        logger.startProgress('Reassembling application');
+    try {
+      await observatory.flutterReassemble(observatory.firstIsolateId);
+    } catch (_) {
+      reassembleStatus.stop(showElapsedTime: true);
+      printError('Reassembling application failed.');
+      return false;
+    }
+    reassembleStatus.stop(showElapsedTime: true);
+    return true;
+  }
+
+  void _printHelp() {
+    printStatus('Type "h" or F1 for this help message. Type "q", F10, or ctrl-c to quit.', emphasis: true);
+    String hot = '';
+    String cold = '';
+    if (hotMode)
+      hot = 'Type "r" or F5 to perform a hot reload of the app';
+    if (device.supportsRestart) {
+      if (hotMode) {
+        cold = ', and "R" to cold restart the app';
+      } else {
+        cold = 'Type "r" or F5 to restart the app';
+      }
+    }
+    if (hot != '' || cold != '')
+      printStatus('$hot$cold.', emphasis: true);
+    printStatus('Type "w" to print the widget hierarchy of the app, and "t" for the render tree.', emphasis: true);
```

Comment:
it's sad that we can't reuse this code more (it's basically identical in two places now).